File: typical90/001/main.py
# -*- coding: utf-8 -*-

n, l = map(int,input().split())
k = int(input())
a = [0] + list(map(int,input().split())) + [l]

def check(l):
    left = 0
    for i in range(k+1):
        right = left
        while(right < len(a) and a[right] - a[left] < l):
            right += 1
        if right == len(a):
            return False
        left = right
    return True

bottom, top = 1, l+1

# 二部探索
while top - bottom > 1:
    mid = (top + bottom) // 2
    if check(mid):
        bottom = mid
    else:
        top = mid

print(bottom)

File: typical90/001/main.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;
#define rep(i, n) for (int i = 0; i < (int)(n); i++)
#define Rep(i,a,b) for(int i=(a);i<(b);i++)
#define all(x) (x).begin(),(x).end()

 ll n, l, k;
 ll a[100005];

bool check(ll m) {
  ll cnt = 0, pre = 0;
  Rep(i, 1, n + 1)
    if (a[i] - pre >= m && l - a[i] >= m) {cnt++; pre = a[i];}
  return cnt >= k;
}

int main() {
  cin >> n >> l >> k;
  Rep(i, 1, n + 1) cin >> a[i];
  ll left = -1, right = l + 1;
  while (right - left > 1) {
    ll mid = (left + right) / 2;
    if (!check(mid)) right = mid;
    else left = mid;
  }
  cout << left << endl;
  return 0;
}

